File: C++/1014. Best Sightseeing Pair.cpp
// O(n)
class Solution {
public:
    int maxScoreSightseeingPair(vector<int>& values) {
        int maxAns = INT_MIN;
        int firstIter = values.size() - 2;
        int maxJ = INT_MIN;
        while(firstIter >= 0) {
            maxJ = max(maxJ, values.at(firstIter + 1) - firstIter - 1);
            maxAns = max(maxAns, values.at(firstIter) + firstIter + maxJ);
            firstIter --;
        }
        return maxAns;
    }
};

// O(n^2)
class Solution2 {
public:
    int maxScoreSightseeingPair(vector<int>& values) {
        int maxAns = INT_MIN;
        for (int firstIter = 0; firstIter < values.size(); firstIter++) {
            for (int secondIter = firstIter + 1; secondIter < values.size(); secondIter++) {
                if (values.at(firstIter) + values.at(secondIter) + firstIter - secondIter > maxAns)
                    maxAns = values.at(firstIter) + values.at(secondIter) + firstIter - secondIter;
            }
        }
        return maxAns;
    }
};

File: Go/1014. Best Sightseeing Pair.go
// O(n)
func maxScoreSightseeingPair(values []int) int {
    var max = math.MinInt64
    var iter = len(values) - 2
    var maxJ = math.MinInt64
    for iter >= 0 {
        if values[iter+1] - (iter+1) > maxJ {
            maxJ = values[iter+1] - (iter+1)
        }
        if values[iter] + iter + maxJ > max {
            max = values[iter] + iter + maxJ
        }  
        iter --      
    }
    return max
}

// O(n^2)
func maxScoreSightseeingPair(values []int) int {
    var max = math.MinInt64
    for firstIter := 0; firstIter < len(values); firstIter++ {
        for secondIter := firstIter + 1; secondIter < len(values); secondIter++ {
            if values[firstIter] + firstIter + values[secondIter] - secondIter > max         {
                max = values[firstIter] + firstIter + values[secondIter] - secondIter
            }
        }
    }
    return max
}